### PR TITLE
Solution 2: Sum of odd numbers

### PR DESCRIPTION
**Pull request by suhrobmuboraksho**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
- [ ] The function has correct type annotations

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by Mithchell509 — APPROVED** [review]

**Comment by Mithchell509** [comment]
Great job Sukhrob

**Review by Mithchell509 — APPROVED** [review]

File: solutions/sum_odd_numbers.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    Returns:
+    int: The sum of all odd numbers in the provided list.
+    None: If the list is invalid.
+
+    Raises:
+    TypeError: If the input is not a list.
+
+    Examples:
+    >>> sum_of_odds([1, 2, 3, 4, 5])
```
Comment (Alemayehu-Desta) [inline solutions/sum_odd_numbers.py:25]: In the example, the function uses "the sum_of_odds" instead of "sum_odd_number" when it is defined and the entire file, just for consistency.

File: solutions/tests/test_sum_odd_numbers.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            str(context.exception),
+            "The function requires a list of integers as an argument.",
+        )
+
+    def test_argument_is_str(self):
+        """Raises a TypeError when the input is a string."""
+        with self.assertRaises(TypeError) as context:
+            sum_odd_numbers("not a list")
+        self.assertEqual(str(context.exception), "Input must be a list of integers.")
+
```
Comment (Alemayehu-Desta) [inline solutions/tests/test_sum_odd_numbers.py:60]: Checking out how the function behaves with the Sum of a very large list of numbers to ensure it runs efficiently would add comprehensiveness. That is applying extreme edge cases.

**Review by Alemayehu-Desta — APPROVED** [review]
Great code Sukhrob,
I learned a lot from your code. It's extensive and easily readable. I forwarded my two cents

File: solutions/sum_odd_numbers.py (hunk 1 above)
Comment (suhrobmuboraksho) [reply solutions/sum_odd_numbers.py:25]: Will make the change. Thanks

**Review by Alemayehu-Desta — APPROVED** [review]

**Review by Mithchell509 — APPROVED** [review]
